### DISCUSSION TITLE
I just made a proof of concept for a Svelte 5 wrapper for StateAdapt

### DISCUSSION THREAD

**Discussion by Evertt** [state-adapt/state-adapt#73] [Ideas]
[Check out this REPL](https://svelte-5-preview.vercel.app/#H4sIAAAAAAAAE41V227jNhD9lamQB2nLpffyZsQBUuxu0aJdB-u0L1HQ6DJy6MikQFJODEP_XvAmybI3uw9B5OGc4ZxzhuQhqliNKprfHSKebTGaR9dNE5FI7xvzQ-2w1hiRSIlWFiZyqQrJGg11xteLNNIqja5Snmq2bYTUcIBCtFyjJMB4IXGLXBOQqFBDB5UUW0gjOvNJdKPSKOWXM1fUFOKXzZVdncPBZ3XwBj7Aov9NS9HmNZbd5axxkLzVWnAQvKhZ8bQ4xAksroYG4o9J55oMEcj38NFs7JDnq9yNKbx9D_euSImjIu9_VMRSd8Bv5nOUH5FoK0pWMSyjuZYtdqR3wVMdnNiosQu92lmZNZrAyi6MFVY60_jWLlPnohc75fhiwYXgSg8UYQEcn32lOI36hYs0SiYg5-cEYIMueZLuycDCtRu_I3A4smMOse2XAIcFvLfu2QD8CpyYVFt8Ds7YdzZk_hTWWGgh1dxVTLWfjbnH94XewAeL6gasbXxADma733ZLh0l5l6T8Nb9O5f6edV6ZwznPbGTqU9XyQjPBvXiMM82y-t-sbpG4GMrEkahRB95wYT-O0pNA3dki8o05VY7tOkDjJISMBLqV3MVdzMunVZ_O8XllPkaw0EJYOoZqsdKS8fV4o9kMbh-ZAqZAC8gRsrxG8ymxQom8wFGmfkTTOxYaSiax0PUeGLdhjdumNoM0Ske6pvDQXyYPwLjSmJUgqiFMbc_dw4T5w8XhaKEL8zDW0bPw82QkHVTy4zkOUkolZtZSZauFc4HSuJ7qSkiIjZVPaIn12b1gIt_cPeH-HhYQU0ozuVb2YJw40ENtejhkAeI5BUpntu4J9Fs7yiFu5qxEyXZYxn3uaK8kcXVTvbSG0RIrxvFGigal3sci3xCzFxlaX5tzHmZ-PCPDOPqNwkLneSSvEXHmJBOzbKtUtbl5gHKMnVpHUhrCokZai3X852r5lSo7vaza--wkCanBFuuJXaNNtq9FViZnWvRsRL5Jeffq3eKujh89BavWnYn-QpEv0xu_zpQK9w6-aOSl6mGWsDGNFfDp85frf_66_W95c_vH8utquCXYmguJn3fItZqD6ZOMGNnZkK3xJja9EhCNnT47_oP2bYMy9pLoR6aoSYYFmH-j6AgcFKaUuvbppEUyyvA4Moz3aHIr7g9NcMaYbbfj-KLHC-EVsC758VWob6TQwnS6rOKKE4sd2zotE5qfzeC3PZRYZW2tCTwjODHh0_JvQKuoGVQPU17vCuKxFnRswOhseBQsAp5WrNYo452h90u8s5dexgsUFVh0Mj797jZLtZ8F2gSOVhRaZHVt2yBwsCbNobeNhB2hH24rxGwGv6O2l3JfDYrHjHE3nn4KtegHUAtY5grlztz9KTeH1yX1isMi-LA-9sHPRL9R4uGu8s_ibybwoZufqDDZKvEaXJel1eBLeMQHMTKnDnL7Ep0RyhXgJWyzJwTVSitXqxAK0TBUJzD7eHKhYYtKwTPTj-6tlGzNeFarlL9KaTLcB_NOnAN0pKdzJPn35T5X-SS7I-cUT0zT54pMPCMnbiend-p99z8QH_-v8QwAAA==)

It's not really a proof of concept for a real wrapper, but more like a proof of concept for what kind of API is possible to create.

## Highlights

* A `Source` is callable, so in the template you can write `onclick={reset}`.
* You can even do `onclick={[increment,3]}`, which will make sure `increment` is always called with `3` as the payload.
* In the template, you can refer to `{counter}` directly, instead of needing to write `{counter.state}`.

As you can see I'm not following the convention of adding a `$` at the end of observables. That's just personal preference, because that's just not common in Svelte land.

**Comment by mfp22** [comment]
Whoa... That's really interesting. I like what you did with `Source`. 

I don't so much like the eager `subscribe` in the `adapt` function. I planned on creating for Angular (and eventually Svelte) another function for when you actually consume a store's state. So `adapt` creates the store, `useStore` subscribes, and `useAdapt` does both at the same time. 

Sometimes I switch between `Source` and `Subject`... Maybe I should stop doing that and just stick with `Source` and stop having to call `.next()` on it. 

Why do you ignore DOM events by default?

As for naming, I think it's important to still have access to observables of selectors of stores, and not just expose signals. So there is a naming challenge here. I did a Twitter poll and decided the best thing would be to have observable selectors as `state$` and signal selectors as `$state`. Yeah, that's awkward with Svelte, but I tested it and it technically works. 

So by default you can put `{ store } ` in  the template and that's cool, but can you still use a different selector, like `{ store.$double }` if you wanted to? I'm assuming yes but it's late for me and I'm too tired to really dig into the details of your implementation more than I have.

Thanks for doing this.

**Comment by Evertt** [comment]
> I don't so much like the eager `subscribe` in the `adapt` function. ... So `adapt` creates the store, `useStore` subscribes, and `useAdapt` does both at the same time.

Yes, agreed. I just realized that and wanted to change it and then I saw your comment. So yeah, as long as the developer using this library knows that `useAdapt` is for actually using it inside a component, then `useAdapt` can eagerly subscribe and also set itself up to automatically unsubscribe whenever the component gets destroyed.

> Why do you ignore DOM events by default?

Well, because in my adapter, I have this state change function for example:

```js
{
    increment: (state, n = 1) => state + n,
}
```

So `increment` can be called with or without a payload and `n` will either be `1` or whatever the payload was. That means that if the template says `onclick={increment}`, that `increment` will be called with a DOM event as the payload. If the `Source` doesn't automatically ignore DOM events in the payload, then you can't use this pretty syntax.

Then you'd either need to change the call site, like `onclick={() => increment(1)}` (because then `onclick={[increment,1]}` also wouldn't work anymore). Or you'd need to change the state change function like so: `increment: (state, n) => state + (+n || 1)`. Both options result in uglier code. And at least in my experience I rarely need information from the DOM event in my state change functions.

> decided the best thing would be to have observable selectors as `state$` and signal selectors as `$state`.

Oof, at first sight that makes me shudder. I'll see if I can get used to that, but if not then I'll just write my own wrapper that creates the naming that I prefer. Cause I think I'd prefer to use whatever naming conventions already exist in Svelte land, which means `$state` for observables and `state` for signals. Of course I understand that outside of Svelte land that doesn't apply. And I get it if you'd prefer to just use a single naming convention everywhere instead of different ones for different frameworks. So yeah, I'll just write my own wrapper if I can't get used to your preferred naming conventions. 😁

> So by default you can put `{ store }`  in the template and that's cool, but can you still use a different selector, like `{ store.$double }` if you wanted to?

Yes yes, you can actually see it in my example code. I'm using both `{ counter }` and `{ counter.double }`. The trick is very simple. Svelte converts any value in the template to a string, so I simply implemented a `toString()` method to the object that is returned by `adapt()`.
